### PR TITLE
chore(metadata): #908 Allow partial inserts to MetadataStore

### PR DESCRIPTION
**Pull request by k88hudson**
This change is [Reviewable](https://reviewable.io/reviews/mozilla/activity-stream/909)

### CODE REVIEW

File: lib/MetadataStore.js [hunk 1]
```diff
@@ -231,39 +237,44 @@ MetadataStore.prototype = {
 
         /* 2. inserts the favicon into page_images, handles the special case, in which
          * the favicon already exsits */
-        let faviconBindings = this._getFaviconParameters(metaObject);
```
Comment (k88hudson) [inline lib/MetadataStore.js]: This is just nesting the block inside an `if (metaObject.favicon_url) {`, sorry the diff is so bad

File: lib/MetadataStore.js [hunk 2]
```diff
@@ -231,39 +237,44 @@ MetadataStore.prototype = {
...
               throw e; /* force this transaction to rollback */
             }
           }
         }
 
-        /* 4. inserts relations into the page_metadata_images */
-        for (let image_id of image_ids) {
-          yield this._conn.executeCached(SQL_INSERT_METADATA_IMAGES, [metadata_id, image_id]);
+        /* 3. inserts all the preview images, if they exist */
+        if (metaObject.images) {
```
Comment (k88hudson) [inline lib/MetadataStore.js]: same thing here

File: test/test-MetadataStore.js [hunk 3]
```diff
@@ -68,12 +68,69 @@ exports.test_insert_single = function*(assert) {
...
+ */
+exports.test_insert_required_fields = function*(assert) {
+  const correctMessage = "Objects to insert must include a places_url";
+  let error;
+  try {
+    yield gMetadataStore.asyncInsert([{}]);
+  } catch (e) {
+    error = e;
+  }
+  assert.ok(error, "Error was thrown for missing cache_key");
```
Comment (k88hudson) [inline test/test-MetadataStore.js]: oops, good catch

File: lib/MetadataStore.js [hunk 4]
```diff
@@ -231,39 +237,44 @@ MetadataStore.prototype = {
...
+          for (let image of metaObject.images) {
+            let imageBindings = this._getImageParameters(image);
+            try {
+              yield this._conn.executeCached(SQL_INSERT_IMAGES, imageBindings);
+              image_ids.push(yield this._asyncGetLastInsertRowID());
+            } catch (e) {
+              try {
+                image_ids.push(yield this._asyncGetImageIDByURL(imageBindings.url));
+              } catch (e) {
+                Cu.reportError("MetadataStore failed to fetch the id of favicon: " + e.message);
```
Comment (ncloudioj) [inline lib/MetadataStore.js]: Oops, found a typo of my previous commit, it should be "MetadataStore failed to fetch the id of **image:** ".
Could you fix this for me, please?

File: lib/MetadataStore.js (hunk 1 above)
Comment (ncloudioj) [inline lib/MetadataStore.js]: 😵 💫 😵 

but good fix 👍

File: lib/MetadataStore.js (hunk 4 above)
Comment (k88hudson) [inline lib/MetadataStore.js]: yep!

**Pull request merged**